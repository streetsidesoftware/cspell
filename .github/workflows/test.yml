name: Test
on:
  pull_request:
  push:
    branches:
      - master
      - cspell4

jobs:
  test:

    runs-on: ${{ matrix.os }}

    strategy:
      matrix:
        node-version:
          - 10.x
          - 12.x
          - 14.x

        os:
          - ubuntu-latest
          - windows-latest
          - macos-latest

    steps:
    - uses: actions/checkout@v2

    - name: Use Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v2.1.2
      with:
        node-version: ${{ matrix.node-version }}

    - run: npm ci
    - run: npm test
    - run: node ./bin.js trace test
    - run: node ./bin.js "**/{*.ts,*.md}"
