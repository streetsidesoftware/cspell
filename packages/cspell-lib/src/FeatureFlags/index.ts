export { FeatureFlag, FeatureFlags, getSystemFeatureFlags, UnknownFeatureFlagError } from './FeatureFlags';
