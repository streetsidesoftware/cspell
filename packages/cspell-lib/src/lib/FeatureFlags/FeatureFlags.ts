export interface FeatureFlag {
    name: string;
    description: string;
}

let systemFeatureFlags: FeatureFlags | undefined;

type FlagTypes = string | boolean;

/**
 * Feature Flags are used to turn on/off features.
 * These are primarily used before a feature has been fully released.
 */
export class FeatureFlags {
    private flags: Map<string, FeatureFlag>;
    private flagValues = new Map<string, FlagTypes>();

    constructor(flags: FeatureFlag[] = []) {
        this.flags = new Map(flags.map((f) => [f.name, f]));
    }

    register(flag: FeatureFlag): this;
    register(name: string, description: string): this;
    register(flagOrName: string | FeatureFlag, description?: string): this {
        if (typeof flagOrName === 'string') {
            return this.register({ name: flagOrName, description: description || '' });
        }
        this.flags.set(flagOrName.name, flagOrName);
        return this;
    }

    getFlag(flag: string): FlagTypes | undefined {
        return this.flagValues.get(flag);
    }

    getFlagBool(flag: string): boolean | undefined {
        return toBool(this.getFlag(flag));
    }

    setFlag(flag: string, value: FlagTypes = true): this {
        if (!this.flags.has(flag)) {
            throw new UnknownFeatureFlagError(flag);
        }

        this.flagValues.set(flag, value);
        return this;
    }

    getFlagInfo(flag: string): FeatureFlag | undefined {
        return this.flags.get(flag);
    }

    getFlags(): FeatureFlag[] {
        return [...this.flags.values()];
    }

    getFlagValues(): Map<string, FlagTypes> {
        return new Map(this.flagValues);
    }

    reset(): this {
        this.flagValues.clear();
        return this;
    }
}

export class UnknownFeatureFlagError extends Error {
    constructor(readonly flag: string) {
        super(`Unknown feature flag: ${flag}`);
    }
}

export function getSystemFeatureFlags(): FeatureFlags {
    return systemFeatureFlags || (systemFeatureFlags = new FeatureFlags());
}

const boolValues: Record<string, boolean | undefined> = {
    0: false,
    1: true,
    f: false,
    false: false,
    n: false,
    no: false,
    t: true,
    true: true,
    y: true,
    yes: true,
};

function toBool(value: boolean | string | undefined): boolean | undefined {
    if (typeof value !== 'string') return value;
    return boolValues[value.toLowerCase()];
}
