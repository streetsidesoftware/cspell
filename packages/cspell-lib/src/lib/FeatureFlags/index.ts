export type { FeatureFlag } from './FeatureFlags.js';
export { FeatureFlags, getSystemFeatureFlags, UnknownFeatureFlagError } from './FeatureFlags.js';
