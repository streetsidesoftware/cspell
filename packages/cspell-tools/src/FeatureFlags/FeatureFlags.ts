export interface FeatureFlag {
    name: string;
    description: string;
}

let systemFeatureFlags: FeatureFlags | undefined;

type FlagTypes = string | boolean;

/**
 * Feature Flags are used to turn on/off features.
 * These are primarily used before a feature has been fully released.
 */
export class FeatureFlags {
    private flags: Map<string, FeatureFlag>;
    private flagValues = new Map<string, FlagTypes>();

    constructor(flags: FeatureFlag[] = []) {
        this.flags = new Map(flags.map((f) => [f.name, f]));
    }

    register(flag: FeatureFlag): this;
    register(name: string, description: string): this;
    register(flagOrName: string | FeatureFlag, description?: string): this {
        if (typeof flagOrName === 'string') {
            return this.register({ name: flagOrName, description: description || '' });
        }
        this.flags.set(flagOrName.name, flagOrName);
        return this;
    }

    registerFeatures(flags: FeatureFlag[]): this {
        flags.forEach((flag) => this.register(flag));
        return this;
    }

    getFlag(flag: string): FlagTypes | undefined {
        return this.flagValues.get(flag);
    }

    getFlagBool(flag: string): boolean | undefined {
        return toBool(this.getFlag(flag));
    }

    setFlag(flag: string, value: FlagTypes = true): this {
        if (!this.flags.has(flag)) {
            throw new UnknownFeatureFlagError(flag);
        }

        this.flagValues.set(flag, toBool(value) ?? value);
        return this;
    }

    getFlagInfo(flag: string): FeatureFlag | undefined {
        return this.flags.get(flag);
    }

    getFlags(): FeatureFlag[] {
        return [...this.flags.values()];
    }

    getFlagValues(): Map<string, FlagTypes> {
        return new Map(this.flagValues);
    }

    reset(): this {
        this.flagValues.clear();
        return this;
    }

    help(): string {
        const flags = [{ name: 'Name', description: 'Description' }, ...this.flags.values()].sort((a, b) =>
            a.name < b.name ? -1 : 1
        );

        const nameColWidth = flags.map((f) => f.name.length).reduce((a, b) => Math.max(a, b), 0) + 1;
        const entries = flags.map((f) => `- ${f.name}${' '.repeat(nameColWidth - f.name.length)} ${f.description}`);

        const text = `Valid Flags:\n${entries.join('\n')}`;
        return text;
    }

    fork(): FeatureFlags {
        const fork = new FeatureFlags([...this.flags.values()]);
        for (const [key, value] of this.flagValues) {
            fork.flagValues.set(key, value);
        }
        return fork;
    }
}

export class UnknownFeatureFlagError extends Error {
    constructor(readonly flag: string) {
        super(`Unknown feature flag: ${flag}`);
    }
}

export function getSystemFeatureFlags(): FeatureFlags {
    return systemFeatureFlags || (systemFeatureFlags = createFeatureFlags());
}

export function createFeatureFlags(): FeatureFlags {
    return new FeatureFlags();
}

const boolValues: Record<string, boolean | undefined> = {
    0: false,
    1: true,
    f: false,
    false: false,
    n: false,
    no: false,
    t: true,
    true: true,
    y: true,
    yes: true,
};

function toBool(value: boolean | string | undefined): boolean | undefined {
    if (typeof value !== 'string') return value;
    return boolValues[value.toLowerCase()];
}
