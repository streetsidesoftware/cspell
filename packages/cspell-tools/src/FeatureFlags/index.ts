export { FeatureFlag, FeatureFlags, getSystemFeatureFlags, UnknownFeatureFlagError } from './FeatureFlags';
export { parseFlags } from './parseFlags';
