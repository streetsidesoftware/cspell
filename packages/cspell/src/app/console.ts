import type { WriteStream } from 'node:tty';
import { formatWithOptions } from 'node:util';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Log = (format?: any, ...params: any[]) => void;

type IOStream = NodeJS.WritableStream &
    Pick<WriteStream, 'isTTY' | 'rows' | 'columns'> &
    Pick<Partial<WriteStream>, 'hasColors' | 'clearLine' | 'getColorDepth'>;

export interface IConsole {
    readonly log: Log;
    readonly error: Log;
    readonly info: Log;
    readonly warn: Log;
    readonly stderrChannel: Channel;
    readonly stdoutChannel: Channel;
    // readonly stderr: IOStream;
    // readonly stdout: IOStream;
}

class ImplChannel implements Channel {
    constructor(readonly stream: IOStream) {}

    write = (msg: string) => this.stream.write(msg);
    writeLine = (msg: string) => this.write(msg + '\n');
    clearLine = (dir: -1 | 0 | 1, callback?: () => void) => this.stream.clearLine?.(dir, callback) ?? false;
    printLine = (...params: unknown[]) =>
        this.writeLine((params.length && formatWithOptions({ colors: this.stream.hasColors?.() }, ...params)) || '');
    getColorLevel = () => getColorLevel(this.stream);
}

class Console implements IConsole {
    readonly stderrChannel: Channel;
    readonly stdoutChannel: Channel;
    constructor(
        readonly stdout = process.stdout,
        readonly stderr = process.stderr,
    ) {
        this.stderrChannel = new ImplChannel(this.stderr);
        this.stdoutChannel = new ImplChannel(this.stdout);
    }

    log: Log = (...p): void => this.stdoutChannel.printLine(...p);
    error: Log = (...p) => this.stderrChannel.printLine(...p);

    info: Log = this.log;
    warn: Log = this.error;
}

export const console: IConsole = new Console();

export function log(...p: Parameters<typeof console.log>) {
    console.log(...p);
}

export function error(...p: Parameters<typeof console.error>) {
    console.error(...p);
}

export interface Channel {
    stream: IOStream;
    write: (msg: string) => void;
    writeLine: (msg: string) => void;
    clearLine: (dir: -1 | 0 | 1, callback?: () => void) => boolean;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    printLine: (format?: any, ...params: any[]) => void;
    getColorLevel: () => 0 | 1 | 2 | 3;
}

export function getColorLevel(stream: IOStream): 0 | 1 | 2 | 3 {
    const depth = stream.getColorDepth?.() || 0;
    switch (depth) {
        case 1: {
            return 1;
        }
        case 4: {
            return 2;
        }
        case 24: {
            return 3;
        }
        default: {
            return 0;
        }
    }
}
