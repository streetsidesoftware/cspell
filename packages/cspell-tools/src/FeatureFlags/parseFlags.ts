import type { FeatureFlags } from './FeatureFlags.js';
import { UnknownFeatureFlagError } from './FeatureFlags.js';

const splitFlag = /[:=]/;
const leadingEql = /^=/;

export function parseFlags(ff: FeatureFlags, flags: string[]): FeatureFlags {
    for (const flag of flags) {
        const [name, value] = flag.replace(leadingEql, '').split(splitFlag, 2);

        try {
            ff.setFlag(name, value);
        } catch (e) {
            if (e instanceof UnknownFeatureFlagError) {
                console.error(e.message);
                console.error(ff.help());
            }
            throw e;
        }
    }

    return ff;
}
