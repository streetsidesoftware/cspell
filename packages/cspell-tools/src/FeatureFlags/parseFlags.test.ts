import { parseFlags } from './parseFlags';
import type { FeatureFlag } from './FeatureFlags';
import { createFeatureFlags } from './FeatureFlags';
import { spyOnConsole } from '../test/console';

const { consoleOutput } = spyOnConsole();

describe('parseFlags', () => {
    const features: FeatureFlag[] = [mkFeature('flag1'), mkFeature('compress'), mkFeature('advanced')];

    beforeEach(() => {
        jest.resetAllMocks();
    });

    test.each`
        flags
        ${[]}
        ${['compress', 'advanced:false', '=flag1']}
        ${['compress:level1', 'advanced:yes', 'flag1=no']}
        ${['compress:"yes"']}
    `('parseFlags', ({ flags }) => {
        const ff = createFeatureFlags();
        ff.registerFeatures(features);
        parseFlags(ff, flags);
        expect(ff.getFlagValues()).toMatchSnapshot();
    });

    test('unknown flag', () => {
        const ff = createFeatureFlags();
        ff.registerFeatures(features);
        expect(() => parseFlags(ff, ['mistaken:true'])).toThrow();
        expect(consoleOutput()).toMatchSnapshot();
    });
});

function mkFeature(name: string, description?: string): FeatureFlag {
    return {
        name,
        description: description ?? name + ' desc',
    };
}
