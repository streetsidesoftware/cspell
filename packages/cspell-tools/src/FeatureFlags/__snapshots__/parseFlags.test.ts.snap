// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseFlags parseFlags 1`] = `Map {}`;

exports[`parseFlags parseFlags 2`] = `
Map {
  "compress" => true,
  "advanced" => false,
  "flag1" => true,
}
`;

exports[`parseFlags parseFlags 3`] = `
Map {
  "compress" => "level1",
  "advanced" => true,
  "flag1" => false,
}
`;

exports[`parseFlags parseFlags 4`] = `
Map {
  "compress" => ""yes"",
}
`;

exports[`parseFlags unknown flag 1`] = `
{
  "error": "Unknown feature flag: mistaken
Valid Flags:
- Name      Description
- advanced  advanced desc
- compress  compress desc
- flag1     flag1 desc",
  "log": "",
}
`;
